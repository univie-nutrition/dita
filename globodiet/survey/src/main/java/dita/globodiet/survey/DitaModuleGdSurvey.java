/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * 
 *       http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
// Auto-generated by Causeway-Stuff code generator.
package dita.globodiet.survey;

import dita.globodiet.survey.dom.Campaign;
import dita.globodiet.survey.dom.CampaignDeps;
import dita.globodiet.survey.dom.Campaign_delete;
import dita.globodiet.survey.dom.Campaign_survey;
import dita.globodiet.survey.dom.Survey;
import dita.globodiet.survey.dom.SurveyDeps;
import dita.globodiet.survey.dom.Survey_delete;
import java.lang.Class;
import java.lang.String;
import javax.annotation.processing.Generated;
import org.apache.causeway.commons.collections.Can;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Import;

@Generated("io.github.causewaystuff.companion.codegen.domgen._GenModule")
@Configuration
@Import({

        // Config Beans
        ModuleConfig.class,

        // Menu Entries
        EntitiesMenu.class,

        // Entities
        Campaign.class,
        Survey.class,

        // Submodules
        CampaignDeps.class,
        SurveyDeps.class,

        // Mixins
        Campaign_delete.class,
        Campaign_survey.class,
        Survey_delete.class,
        })
public class DitaModuleGdSurvey {
    public static final String NAMESPACE = "dita.globodiet.survey";

    public static Can<Class<?>> entityClasses() {
        return Can.of(Campaign.class,
        Survey.class);
    }
}
