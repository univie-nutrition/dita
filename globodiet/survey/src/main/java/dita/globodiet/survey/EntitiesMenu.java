/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * 
 *       http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
// Auto-generated by Causeway-Stuff code generator.
package dita.globodiet.survey;

import dita.globodiet.survey.dom.Campaign;
import dita.globodiet.survey.dom.ReportColumnDefinition;
import dita.globodiet.survey.dom.RespondentFilter;
import dita.globodiet.survey.dom.Survey;
import io.github.causewaystuff.companion.applib.services.search.SearchService;
import jakarta.inject.Inject;
import jakarta.inject.Named;
import javax.annotation.processing.Generated;
import org.apache.causeway.applib.annotation.Action;
import org.apache.causeway.applib.annotation.ActionLayout;
import org.apache.causeway.applib.annotation.DomainService;
import org.apache.causeway.applib.services.factory.FactoryService;

@Generated("io.github.causewaystuff.companion.codegen.domgen._GenMenu")
@Named("dita.globodiet.survey.EntitiesMenu")
@DomainService
public class EntitiesMenu {
    @Inject
    private FactoryService factoryService;

    @Inject
    private SearchService searchService;

    @Action
    @ActionLayout(
            cssClassFa = "solid users-viewfinder .campaign-color"
    )
    public Campaign.Manager manageCampaign() {
        return factoryService.viewModel(new Campaign.Manager(searchService, ""));
    }

    @Action
    @ActionLayout(
            cssClassFa = "solid table-columns .reportColumnDefinition-color\n"
    )
    public ReportColumnDefinition.Manager manageReportColumnDefinition() {
        return factoryService.viewModel(new ReportColumnDefinition.Manager(searchService, ""));
    }

    @Action
    @ActionLayout(
            cssClassFa = "solid user .respondentFilter-color,\n"
                            + "solid filter .survey-color .ov-size-60 .ov-right-55 .ov-bottom-55\n"
    )
    public RespondentFilter.Manager manageRespondentFilter() {
        return factoryService.viewModel(new RespondentFilter.Manager(searchService, ""));
    }

    @Action
    @ActionLayout(
            cssClassFa = "solid users .survey-color"
    )
    public Survey.Manager manageSurvey() {
        return factoryService.viewModel(new Survey.Manager(searchService, ""));
    }
}
