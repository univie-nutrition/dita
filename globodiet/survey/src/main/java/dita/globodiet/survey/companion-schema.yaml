# -----------------------------------------------------------
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
# 
#       http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
# -----------------------------------------------------------

module:
  namespace: dita.globodiet.survey
  package: dita.globodiet.survey
viewmodels:
entities:
- id: dom.Campaign
  name: Campaign
  namespace: dom
  table: CAMPAIGN
  secondaryKey: |
    SURVEY
    CODE
  title: String.format("%s (code=%s|%s)", name, surveyCode, code)
  icon: solid users-viewfinder .campaign-color
  description: |
    A campaign defines a part of a food consumption survey that contains several interviews.
    Campaigns can be defined to be the whole study, 
    a seasonal part of a study, a regional part of a study etc.
  fields:
    surveyCode:
      column: SURVEY
      column-type: nvarchar(20)
      required: true
      unique: false
      foreignKeys: |
        SURVEY.CODE
      description: |
        Survey code   
    code:
      column: CODE
      column-type: nvarchar(20)
      required: true
      unique: false
      description: |
        Unique (survey scoped) campaign identifier.
    name:
      column: NAME
      column-type: nvarchar(120)
      required: true
      unique: false
      description: |
        Descriptive campaign name.
    description:
      column: DESCRIPTION
      column-type: nvarchar(240)
      required: false
      unique: false
      multiLine: 4
      description: |
        Detailed information for this campaign.
- id: dom.ReportColumnDefinition
  name: ReportColumnDefinition
  namespace: dom
  table: ReportColumnDefinition
  secondaryKey: |
    SURVEY
    NAME
  title: String.format("%s (%s)", name, surveyCode)
  icon: |
    solid table-columns .reportColumnDefinition-color
    
  description: |
    A column-definition defines which columns to include 
    with an interview report.
  fields:
    surveyCode:
      column: SURVEY
      column-type: nvarchar(20)
      required: true
      unique: false
      foreignKeys: |
        SURVEY.CODE
      description: |
        Survey code   
    name:
      column: NAME
      column-type: nvarchar(120)
      required: true
      unique: false
      description: |
        Descriptive column-definition name, unique within survey.
    description:
      column: DESCRIPTION
      column-type: nvarchar(240)
      required: false
      unique: false
      multiLine: 4
      description: |
        Detailed information for this column-definition.
    columnListing:
      column: COLLISTING
      column-type: mediumtext
      required: false
      unique: false
      cssClass: listing
      fieldSet: listing
      multiLine: 24
      hiddenWhere: EVERYWHERE
      labelPosition: NONE
      description: |
        Line by line defines a column to include.
        (lines can be commented out with a leading #)
- id: dom.RespondentFilter
  name: RespondentFilter
  namespace: dom
  table: RespondentFilter
  secondaryKey: |
    SURVEY
    NAME
  title: String.format("%s (%s)", name, surveyCode)
  icon: |
    solid user .respondentFilter-color,
    solid filter .survey-color .ov-size-60 .ov-right-55 .ov-bottom-55
    
  description: |
    A respondent filter defines which respondents to include 
    with an interview report.
  fields:
    surveyCode:
      column: SURVEY
      column-type: nvarchar(20)
      required: true
      unique: false
      foreignKeys: |
        SURVEY.CODE
      description: |
        Survey code   
    name:
      column: NAME
      column-type: nvarchar(120)
      required: true
      unique: false
      description: |
        Descriptive filter name, unique within survey.
    description:
      column: DESCRIPTION
      column-type: nvarchar(240)
      required: false
      unique: false
      multiLine: 4
      description: |
        Detailed information for this filter.
    aliasListing:
      column: ALIAS
      column-type: mediumtext
      required: false
      unique: false
      cssClass: listing
      fieldSet: listing
      multiLine: 24
      hiddenWhere: EVERYWHERE
      labelPosition: NONE
      description: |
        Line by line defines an alias to include.
        (lines can be commented out with a leading #)
- id: dom.Survey
  name: Survey
  namespace: dom
  table: SURVEY
  secondaryKey: |
    CODE
  title: String.format("%s (code=%s)", name, code)
  icon: solid users .survey-color
  description: |
    A survey collects one or more campaigns.
  fields:
    code:
      column: CODE
      column-type: nvarchar(20)
      required: true
      unique: true
      description: |
        Unique (application scoped) survey identifier.
    name:
      column: NAME
      column-type: nvarchar(120)
      required: true
      unique: false
      description: |
        Descriptive survey name.
    systemId:
      column: SYSID
      column-type: nvarchar(20)
      required: true
      unique: false
      description: |
        System ID part of semantic identifiers for this survey.
        e.g. at.gd/2.0
    correction:
      column: CORRECTION
      column-type: clob
      required: false
      unique: false
      hiddenWhere: EVERYWHERE
      description: |
        Yaml formatted interview data corrections.

