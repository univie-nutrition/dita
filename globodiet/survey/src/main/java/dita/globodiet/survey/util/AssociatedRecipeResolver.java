/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package dita.globodiet.survey.util;

import java.math.BigDecimal;
import java.util.Optional;

import org.apache.causeway.commons.collections.Can;
import org.apache.causeway.commons.internal.exceptions._Exceptions;

import lombok.NonNull;

import dita.commons.food.consumption.FoodConsumption.ConsumptionUnit;
import dita.commons.format.FormatUtils;
import dita.commons.sid.SemanticIdentifier;
import dita.commons.sid.SemanticIdentifier.SystemId;
import dita.commons.sid.SemanticIdentifierSet;
import dita.foodon.fdm.FoodDescriptionModel;
import dita.foodon.fdm.FoodDescriptionModel.Recipe;
import dita.recall24.dto.RecallNode24;
import dita.recall24.dto.RecallNode24.Annotation;
import dita.recall24.dto.RecallNode24.Transfomer;
import dita.recall24.dto.Record24;
import dita.recall24.dto.Record24.Composite;
import dita.recall24.dto.Record24.Food;

/**
 * In 2024, the Austrian GloboDiet food description model introduced the concept of 'associated' food,
 * that can be quickly entered during interviews,
 * but requires a post-processing step later (before interview reporting),
 * with which those reported food nodes are to be replaced by an associated recipe.
 * <p>
 * The association link between food and recipe is made via references in their name:
 * <ul>
 * <li>food reference their associated recipe via suffix {assocRecp=‹recipe-code›}</li>
 * <li>recipes reference their associated food via suffix {assocFood=‹food-code›}</li>
 * </ul>
 */
public record AssociatedRecipeResolver(
        @NonNull FoodDescriptionModel foodDescriptionModel
        ) implements Transfomer {

    @SuppressWarnings("unchecked")
    @Override
    public <T extends RecallNode24> T transform(final T node) {

        return switch(node) {
            // we replace the original food by its associated recipe
            case Food origFood -> {

                var fdmAdapter = new FDMAdapter(origFood.sid().systemId(), foodDescriptionModel);

                var associatedRecipe = fdmAdapter
                        .associatedRecipe(NameWithCode.parseAssocRecipe(origFood.name()))
                        .orElse(null);
                if(associatedRecipe==null) yield (T)origFood;

                var recordBuilder = new Composite.Builder();

                // replace the (proxy-) food node by its associated composite node ..

                recordBuilder.type(Record24.Type.COMPOSITE);
                recordBuilder.name(NameWithCode.parseAssocFood(associatedRecipe.name()).nameWithResolvedSuffix());
                recordBuilder.sid(associatedRecipe.sid());
                // keep the original food as comment
                recordBuilder.subRecords().add(origFoodAsComment(origFood));

                // store GloboDiet food description group data as annotation
                recordBuilder.annotations().clear();
                recordBuilder.annotations().add(new Annotation("group", recipeGroupSid(associatedRecipe)));

                //TODO[dita-globodiet-survey-24] what facets to put here?
                recordBuilder.facetSids(SemanticIdentifierSet.wip());

                var origFoodConsumedOverRecipeMass = new BigDecimal(
                        origFood.amountConsumed().doubleValue() //TODO[dita-globodiet-survey] might not always be in GRAM
                        / foodDescriptionModel.sumAmountGramsForRecipe(associatedRecipe).doubleValue());

                // to the composite record we are building here,
                // we add each recipe ingredient as sub-record
                foodDescriptionModel.streamIngredients(associatedRecipe)
                    .map(ingr->{
                        var food = foodDescriptionModel.lookupFoodBySid(ingr.foodSid()).orElseThrow();
                        var foodBuilder = new Food.Builder()
                            .name(food.name())
                            .sid(ingr.foodSid())
                            .facetSids(SemanticIdentifierSet.wip()) //TODO[dita-globodiet-survey] missing ingredient facets
                            .amountConsumed(ingr.amountGrams().multiply(origFoodConsumedOverRecipeMass))
                            .consumptionUnit(ConsumptionUnit.GRAM); // for recipes this is always in GRAM
                        foodBuilder.annotations().add(new Annotation("group", foodGroupSid(food)));
                        return foodBuilder.build();
                    })
                    .forEach(recordBuilder.subRecords()::add);

                yield (T)recordBuilder.build();

            }
            default -> node;
        };
    }

    // -- HELPER

    private SemanticIdentifier foodGroupSid(final FoodDescriptionModel.Food food) {
        return food.groupSid()
                .mapObjectId(o->o.mapContext(_->SidUtils.GdContext.FOOD_GROUP.id()));
    }

    private SemanticIdentifier recipeGroupSid(final FoodDescriptionModel.Recipe recipe) {
        return recipe.groupSid()
                .mapObjectId(o->o.mapContext(_->SidUtils.GdContext.RECIPE_GROUP.id()));
    }

    private Record24.Comment origFoodAsComment(final Food origFood) {
        var name = "%s, amount=%s, raw/cooked=%s".formatted(
                origFood.name().replace(" {", ", ").replace("}", ""),
                origFood.consumptionUnit().format(origFood.amountConsumed()),
                origFood.rawPerCookedRatio());

        return Record24.comment(name, origFood.sid(), origFood.facetSids(),
                Can.ofCollection(origFood.annotations().values()));
    }

    private record FDMAdapter(
            @NonNull SystemId systemId,
            @NonNull FoodDescriptionModel foodDescriptionModel) {

        /**
         * Optionally returns the associated {@link Recipe},
         * based on whether a suffix {assocRecp=‹recipe-code›} is present.
         * <p>
         * Fails if the recipe-code is present but cannot be resolved.
         */
        Optional<Recipe> associatedRecipe(@NonNull final NameWithCode nameWithCode) {
            var associatedRecipeSid = nameWithCode.associatedRecipeSid(systemId).orElse(null);
            if(associatedRecipeSid == null) return Optional.empty();
            var associatedRecipe = foodDescriptionModel.lookupRecipeBySid(associatedRecipeSid);
            if(associatedRecipe.isEmpty()) {
                throw _Exceptions.illegalArgument("failed to resolve recipe for %s using sid %s",
                        nameWithCode.name(),
                        associatedRecipeSid);
            }
            return associatedRecipe;
        }

    }

    private record NameWithCode(String name, String code) {
        static NameWithCode parseAssocFood(final String nameAndCode) {
            return parse(nameAndCode, "{assocFood=");
        }
        static NameWithCode parseAssocRecipe(final String nameAndCode) {
            return parse(nameAndCode, "{assocRecp=");
        }
        static NameWithCode parse(final String nameAndCode, final String magic) {
            int c1 =nameAndCode.indexOf(magic);
            if(c1==-1) return new NameWithCode(nameAndCode, null);
            int c2 = nameAndCode.indexOf("}", c1);
            return new NameWithCode(
                    nameAndCode.substring(0, c1).trim(),
                    FormatUtils.fillWithLeadingZeros(5, nameAndCode.substring(c1 + magic.length(), c2)));
        }
        String nameWithResolvedSuffix() {
            return name() + " {resolved}";
        }
        private Optional<SemanticIdentifier> associatedRecipeSid(final SystemId systemId) {
            return Optional.ofNullable(code())
                    .map(code->SidUtils.GdContext.RECIPE.sid(systemId, code));
        }
    }

}
