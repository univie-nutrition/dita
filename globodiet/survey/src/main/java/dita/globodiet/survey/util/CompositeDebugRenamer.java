/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package dita.globodiet.survey.util;

import java.util.concurrent.atomic.LongAdder;

import dita.recall24.dto.RecallNode24;
import dita.recall24.dto.RecallNode24.Transfomer;
import dita.recall24.dto.Record24;
import dita.recall24.dto.Record24.Composite;

/**
 * Adds a composite number of occurrence.
 *
 * <p>Phase 2 processes all the candidates as collected from phase 1.
 */
public record CompositeDebugRenamer() implements Transfomer {

    private static final LongAdder c = new LongAdder();

    @SuppressWarnings("unchecked")
    @Override
    public <T extends RecallNode24> T transform(final T node) {
        return switch (node) {
            case Record24.Composite composite -> {
                c.increment();
                var builder = ((Composite.Builder) composite.asBuilder())
                        .name(composite.name() + "(" + c.longValue() + ")");
                builder.replaceSubRecords(this::transform); // recursive
                yield (T) builder.build();
            }
            default -> node;
        };
    }

}
