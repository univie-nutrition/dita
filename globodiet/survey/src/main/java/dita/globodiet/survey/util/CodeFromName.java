/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package dita.globodiet.survey.util;

import java.util.Optional;

import org.jspecify.annotations.Nullable;

import org.springframework.util.StringUtils;

import org.apache.causeway.commons.internal.base._Strings;
import org.apache.causeway.commons.io.TextUtils;

import dita.commons.sid.SemanticIdentifier;
import dita.commons.sid.SemanticIdentifier.ObjectId;
import dita.commons.sid.SemanticIdentifier.SystemId;
import dita.commons.util.FormatUtils;

record CodeFromName(@Nullable String code) {

    static CodeFromName parseAssocFood(final String nameAndCode) {
        return parse("assocFood", nameAndCode);
    }
    static CodeFromName parseAssocRecipe(final String nameAndCode) {
        return parse("assocRecp", nameAndCode);
    }
    private static CodeFromName parse(final String attributeName, final String nameAndCode) {
        int c1 = nameAndCode.indexOf("{");
        return (c1==-1)
            ? new CodeFromName(null)
            : new CodeFromName(
                lookupAttribute(attributeName, nameAndCode)
                    .map(code->FormatUtils.fillWithLeadingZeros(5, code))
                    .orElse(null));
    }
    Optional<SemanticIdentifier> associatedRecipeSid(final SystemId systemId) {
        return Optional.ofNullable(code())
                .map(code->ObjectId.Context.RECIPE.sid(systemId, code));
    }

    /// looks for `value` in `{.., key=value,..}`
    private static Optional<String> lookupAttribute(final String attributeKey, final String curlybracedAttributeList) {
        final String commaSeparatedKeyValuePairs = TextUtils.cutter(curlybracedAttributeList)
            .keepAfter("{")
            .keepBeforeLast("}")
            .getValue();
        final String pattern = attributeKey + "=";

        return _Strings.splitThenStream(commaSeparatedKeyValuePairs, ",")
            .map(String::trim)
            .filter(kv->kv.startsWith(pattern))
            .map(kv->kv.substring(pattern.length()))
            .filter(StringUtils::hasLength)
            .findFirst();
    }

}

