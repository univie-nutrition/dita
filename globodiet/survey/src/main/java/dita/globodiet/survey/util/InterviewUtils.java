/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package dita.globodiet.survey.util;

import java.nio.charset.StandardCharsets;
import java.util.Optional;
import java.util.stream.Stream;

import org.apache.causeway.applib.value.Blob;
import org.apache.causeway.applib.value.Clob;
import org.apache.causeway.applib.value.NamedWithMimeType.CommonMimeType;
import org.apache.causeway.commons.io.ZipUtils.ZipOptions;

import lombok.experimental.UtilityClass;
import lombok.extern.log4j.Log4j2;

import io.github.causewaystuff.blobstore.applib.BlobDescriptor;
import io.github.causewaystuff.blobstore.applib.BlobStore;
import io.github.causewaystuff.commons.base.types.NamedPath;

@UtilityClass
@Log4j2
public class InterviewUtils {

    public Clob unzip(final Blob zippedInterviewSource) {
        var unzippedBlob = zippedInterviewSource
                .unZip(CommonMimeType.XML, ZipOptions.builder()
                        .zipEntryCharset(StandardCharsets.ISO_8859_1)
                        .zipEntryFilter(entry->{
                            log.info(String.format("parsing Zip entry %s (%.2fKB)",
                                    entry.getName(),
                                    0.001*entry.getSize()));
                            return true;
                        })
                        .build());
        return unzippedBlob
                .toClob(StandardCharsets.UTF_8);
    }

    public Stream<Clob> streamSources(final BlobStore surveyBlobStore, final NamedPath path, final boolean recursive) {
        return surveyBlobStore.listDescriptors(path, recursive)
            .stream()
            .filter(desc->desc.mimeType().equals(CommonMimeType.XML))
            .map(BlobDescriptor::path)
            .map(surveyBlobStore::lookupBlob)
            .map(Optional::orElseThrow)
            .map(InterviewUtils::unzip);
    }

}
