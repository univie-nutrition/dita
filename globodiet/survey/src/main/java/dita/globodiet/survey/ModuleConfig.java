/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
// Auto-generated by DitA-Tooling
package dita.globodiet.survey;

import org.causewaystuff.blobstore.applib.BlobStore;
import org.causewaystuff.blobstore.applib.BlobStoreFactory;
import org.causewaystuff.commons.base.types.NamedPath;

import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Import;

import dita.globodiet.survey.entities.SurveyManager_addSurvey;
import dita.globodiet.survey.entities.Survey_addCampaign;
import dita.globodiet.survey.util.InterviewUtils;
import dita.globodiet.survey.view.SurveyTreeNode;
import dita.globodiet.survey.view.SurveyTreeNodeFactory;
import dita.globodiet.survey.view.SurveyVM;
import dita.recall24.model.InterviewSet24;

@Configuration
@Import({
    SurveyVM.class,
    BlobStoreFactory.class,
    
    // non-gen mixins
    SurveyManager_addSurvey.class,
    Survey_addCampaign.class,
        })
@EnableConfigurationProperties({ModuleConfig.SurveyConfiguration.class})
//@ComponentScan(basePackageClasses = Survey.class)
public class ModuleConfig {
    
    @ConfigurationProperties(DitaModuleGdSurvey.NAMESPACE)
    public static record SurveyConfiguration(
            BlobStoreFactory.BlobStoreConfiguration blobstore) {
    }

    @Bean @Qualifier("survey")
    public BlobStore surveyBlobStore(
            final BlobStoreFactory blobStoreFactory,
            final SurveyConfiguration surveyConfiguration) {
        return blobStoreFactory.createBlobStore(surveyConfiguration.blobstore());
    }

    @Bean
    public SurveyTreeNode surveyTreeRootNode(@Qualifier("survey") final BlobStore surveyBlobStore) {

        if(surveyBlobStore==null) {
            return SurveyTreeNodeFactory.emptyNode();
        }

        return InterviewUtils.streamSources(surveyBlobStore, NamedPath.empty(), true)
            .limit(1)
            .map(InterviewUtils::parse)
            .reduce(InterviewSet24::join)
            .map(InterviewSet24::normalized)
            .map(SurveyTreeNodeFactory::surveyNode)
            .orElseGet(SurveyTreeNodeFactory::emptyNode);
    }

}
