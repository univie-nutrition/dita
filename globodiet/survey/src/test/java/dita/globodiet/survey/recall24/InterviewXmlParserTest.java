/*
 *  Licensed to the Apache Software Foundation (ASF) under one
 *  or more contributor license agreements.  See the NOTICE file
 *  distributed with this work for additional information
 *  regarding copyright ownership.  The ASF licenses this file
 *  to you under the Apache License, Version 2.0 (the
 *  "License"); you may not use this file except in compliance
 *  with the License.  You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing,
 *  software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *  KIND, either express or implied.  See the License for the
 *  specific language governing permissions and limitations
 *  under the License.
 */
package dita.globodiet.survey.recall24;

import jakarta.inject.Inject;

import org.causewaystuff.blobstore.applib.BlobStore;
import org.causewaystuff.commons.types.NamedPath;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertNotNull;

import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.boot.test.context.SpringBootTest;

import org.apache.causeway.commons.io.DataSource;

import dita.globodiet.survey.DitaModuleGdSurvey;
import dita.globodiet.survey.PrivateDataTest;
import dita.globodiet.survey.util.InterviewUtils;

@SpringBootTest(classes = {
        DitaModuleGdSurvey.class
        })
@PrivateDataTest
class InterviewXmlParserTest {

    @Inject @Qualifier("survey") BlobStore surveyBlobStore;
    private InterviewXmlParser parser = new InterviewXmlParser();

    @Test
    void parsingFromBlobStore() {
        assertNotNull(surveyBlobStore);

        InterviewUtils.streamSources(surveyBlobStore, NamedPath.empty(), true)
        .limit(1)
        .map(InterviewUtils::parse)
        .forEach(iSet->{
            System.err.printf("%s%n", iSet.toJson());
        });
    }

    @Test
    void parsingSample() {
        var xml = InterviewSampler.sampleXml();
        var iSet = parser.parse(DataSource.ofStringUtf8(xml));
        //System.err.printf("%s%n", iSet.toJson());
    }

}
