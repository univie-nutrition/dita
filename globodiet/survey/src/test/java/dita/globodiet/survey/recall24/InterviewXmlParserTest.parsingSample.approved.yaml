respondents:
- alias: "Svenson|Sven"
  dateOfBirth: "1933-03-17"
  gender: "FEMALE"
  interviews:
  - parentRespondentRef: "ObjectRef[Respondent24]"
    interviewDate: "2021-05-02"
    interviewOrdinalRef:
      value: 1
    respondentMetaData:
      parentInterviewRef: "ObjectRef[Interview24]"
      specialDietId: "99"
      specialDayId: "00"
      heightCM: 160
      weightKG: 54.9
    meals:
    - parentInterviewRef: "ObjectRef[Interview24]"
      hourOfDay: "07:00:00"
      foodConsumptionOccasionId: "01"
      foodConsumptionPlaceId: "01"
      memorizedFood:
      - parentMealRef: "ObjectRef[Meal24]"
        name: "Kaffee"
        records:
        - parentMemorizedFoodRef: "ObjectRef[MemorizedFood24]"
          type: "FOOD"
          name: "Eieraufstrich"
          facetSids: "0103,0200,0317,0405"
          ingredients:
          - parentRecordRef: "ObjectRef[Record24]"
            sid: "01529"
            name: "Mayonnaise"
            facetSids: "0699,0400,1000,0801"
            rawPerCookedRatio: 1.0
            quantityCooked: "6.741572856903076[g]"
          - parentRecordRef: "ObjectRef[Record24]"
            sid: "00916"
            name: "Ei"
            facetSids: "0107,0303"
            rawPerCookedRatio: 1.0
            quantityCooked: "16.853933334350586[g]"
          - parentRecordRef: "ObjectRef[Record24]"
            sid: "01547"
            name: "Schnittlauch"
            facetSids: "0205,0499"
            rawPerCookedRatio: 1.0
            quantityCooked: "0.9363296031951904[g]"
          - parentRecordRef: "ObjectRef[Record24]"
            sid: "00103"
            name: "Knoblauch"
            facetSids: "0205,0499,0999,1200,0399,2800"
            rawPerCookedRatio: 1.0
            quantityCooked: "0.3745318353176117[g]"
          - parentRecordRef: "ObjectRef[Record24]"
            sid: "01571"
            name: "Salz, Koch"
            facetSids: ""
            rawPerCookedRatio: 1.0
            quantityCooked: "0.09363295882940292[g]"
        - parentMemorizedFoodRef: "ObjectRef[MemorizedFood24]"
          type: "FOOD"
          name: "Kaffee, mit Koffein"
          facetSids: "0699,0201,1200,0702"
          ingredients: []
